Sprint 7

Darktel opened the pull request:
Выполнение задания для 7-го спринта. 


Запуск производится командами: 
> `mvn clean test`
> `allure generate target/allure-results -o allure-report --clean`
Просмотр allure отчета 
> `allure open allure-report`

В ходе прогона теста выявлены ошибки которые отображены в аллюр отчете.

Commit:
изменил pom.xml для проекта. подключил зависимости.

Commit:
Добавил тесты и данные для тестов.

Commit:
Добавил allure отчет

Commit:
Добавил allure отчет.
Вынес в отдельные методы создание курьера.
Добавил процесс очистки данных.

pomegranate282922 commented on src/main/java/praktikum/courier/Courier.java:
```diff
@@ -0,0 +1,66 @@
+        return response.jsonPath().getString("id");
+    }
+
+    @Step("Удаление курьера")
```
Такие степы не надо располагать в pojo. Вынеси их пожалуйста в другие классы. 
нужно исправить

Darktel replied:
исправлено перенес в base test классы.

pomegranate282922 commented on src/main/java/praktikum/order/Order.java:
```diff
@@ -0,0 +1,75 @@
+        return color;
+    }
+
+    private final String firstName;
```
Сначала все поля, потом методы 
Можно улучшить

Darktel replied:
Исправлено. [Коммит](https://github.com/Darktel/qa_java/pull/7/commits/058f81919b8a8d4a94c440bb53dad1edb101db40).

pomegranate282922 commented on src/main/java/praktikum/order/Order.java:
```diff
@@ -0,0 +1,75 @@
+        this.color = color;
+    }
+
+
```
Можно использовать библиотеку lombok для уменьшения boilerplate кода 
Можно улучшить

Darktel replied:
Исправлено в этом [коммите](https://github.com/Darktel/qa_java/pull/7/commits/03a9ce94ca87c46b63fd264168b7987b25e64652). 
Но у меня после подключения библиотеки и чистки не нужного кода IDE начала ругаться. Что нет конструктора и есть проблемы в связанном коде. Такое ощущение что IDE не понимает подключение библиотеки lombok. Можно ли как то это исправить?

pomegranate282922 commented on src/test/java/TestCreateCourier.java:
```diff
@@ -0,0 +1,117 @@
+
+    @BeforeEach
+    public void setUp() {
+        RestAssured.baseURI = "https://qa-scooter.praktikum-services.ru/api/v1";
```
Base uri присваиваем один раз в проекте
Нужно исправить

Darktel replied:
А один раз как это лучше сделать ? Т.к. я сейчас остановился на 2-а раза в base test классах. Для теста курьера и тестов по заказам. 
Можно ли оставить так? или реализовать через вызов Callback BeforeAll ? На подобии [этого](https://github.com/Darktel/qa_java/blob/sprint_6/src/main/java/ru/praktikumservices/qascooter/DriverExtension.java).

pomegranate282922 commented on src/test/java/TestCreateCourier.java:
```diff
@@ -0,0 +1,117 @@
+    @DisplayName("Проверка создания не уникального курьера")
+    @Description("Нельзя создать двух одинаковых курьеров")
+    @Test
+    public void TestCreateNewCourierNotUnique() {
```
Методы именуем camelCase 
Нужно исправить

Darktel replied:
Исправлено в этом [коммите](https://github.com/Darktel/qa_java/pull/7/commits/a95cbb395865eb4ac0a31cd5f07aee85a09dc238).

pomegranate282922 commented on src/test/java/TestCreateCourier.java:
```diff
@@ -0,0 +1,117 @@
+    public void TestCreateNewCourierWithoutPasswordFields() {
+        Map<String, String> courier = Map.of("login", "ninja7795", "firstName", "saske");
+        Response response = given()
+                .header("Content-Type", "application/json")
```
Все вызовы ручек выносим в отдельные классы из тестовых классов и переиспользуем 
Нужно исправить

Darktel replied:
Переделано. созданы 2 доп. класса для курьера и заказа. 
[Тут](https://github.com/Darktel/qa_java/pull/7/commits/09a7eb60ede9172f79b7851566abee8ec0118b67) 
и [Тут](https://github.com/Darktel/qa_java/pull/7/commits/9569511cba6681fe2a78c7f9fd3fe14252906501).

pomegranate282922 commented on src/test/java/TestCreateOrder.java:
```diff
@@ -0,0 +1,59 @@
+                Arguments.of("Empty list", Collections.emptyList()),
+                Arguments.of("Null colors", null)
+        );
+    }
```
Отлично применена параметризация

Darktel replied:
Благодарю!

pomegranate282922 commented on src/test/java/TestListOrder.java:
```diff
@@ -0,0 +1,29 @@
+        response.then().assertThat()
+                .statusCode(200)
+                .body("orders",  // в теле ответа есть JSON c объектом orders,
+                hasSize(30)); // содержащий 30 элементов - значение по умолчанию.
```
Всегда 30?

Darktel replied:
Согласен, может не быть заказов вовсе. 
Переделал эту проверку на notNull.

pomegranate282922 commented on src/test/java/TestLoginCourier.java:
```diff
@@ -0,0 +1,107 @@
+    }
+    @AfterEach
+    public void terDown() {
+        if (createdCourierId != null) {
```
Не всегда удаляется созданный курьер 
Нужно исправить

Darktel replied:
В возможных местах если тест получил статус Error. Сработает проверка по status code = 201. Если курьер все же создался. И вызовется его удаление (Именно из зашитых данных в тест).

pomegranate282922 replied:
Не поняла объяснение( 
Прошу перепроверить - курьер создается в каждом (!) тесте, а удаляется - точно в каждом?

pomegranate282922 commented on src/test/resources/newCourierJson.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "login": "ninja999957",
```
Лучше через объект, чем через файл 
Можно улучшить

Darktel replied:
снес файл. переделано на создание через POJO где это возможно.

pomegranate282922 reviewed (commented):
Работа хорошая! Все тесты реализованы. Применена параметризация.
Замечания:
1. Нейминг методов 
2. base Uri присваиваем один раз в проекте 
3. Все вызовы ручек выносим из тестовых классов 
4. Все степы выносим из классов модели
5. Везде обеспечить ударение тестового курьера

Commit:
отрефакторил тесты по созданию и логированию курьера.
Создал отдельный класс BaseCourierTest.java Куда вынес все вызовы API. И в тесте отнаследовался от него.
Так же почистил POJO класc Courier.java

Commit:
отрефакторил тесты по созданию заказа. Все вызовы api перенес в BaseOrderTest.java

Commit:
Убрал не используемые импорты.

Commit:
отрефакторил тест на получение списка заказов.

Commit:
отрефакторил тест на получение списка заказов.
Прогнал через линтер

Commit:
Изменил порядок структуры в POJO файле Order.java

Commit:
подключил библиотеку lombok. переписал файлы Pojo.

Commit:
новый allure отчет.

pomegranate282922 commented on src/main/java/praktikum/base/BaseCourierTest.java:
```diff
@@ -0,0 +1,124 @@
+        response.then()
+                .statusCode(409)
+                .body("message", equalTo("Этот логин уже используется"));
+    }
```
Это тоже не лучшее место. Лучше вообще в отдельный класс.

Darktel replied:
Вынес в отдельные классы. OrderChecks и CourierChecks

pomegranate282922 commented on src/test/java/TestLoginCourier.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Map;
+
+public class TestLoginCourier extends BaseCourierTest {
+    private final BaseCourierTest courierApi = new BaseCourierTest();
```
Это точно нехорошо - и наследуется и поле создаем того же класса? Зачем? Нужно разобраться

Darktel replied:
убрал наследование.

pomegranate282922 reviewed (commented):
Привет! Многое поправил, но еще немного нужно доделать.

Commit:
Вынес конфигурацию спецификации restassured в единое место.
Переделал вспомогательные тестовые классы.

Commit:
поправил gitignor

Commit:
Вынес проверки в отдельные классы.

Commit:
Добавил свежий allure отчет.

Darktel commented:
Я вроде бы со всеми замечаниями разобрался. 
Спасибо что подсказали про плагин для lombok. 
По проверке удаляется ли курьер после каждого теста. Внимательно продебажил. И убрал переменную и проверку ее на null. (осталось после предыдущего рефакторинга... (такой тупой косяк, который я пропустил)).

pomegranate282922 commented on src/test/java/TestLoginCourier.java:
```diff
@@ -0,0 +1,63 @@
+    }
+    @AfterEach
+    public void terDown() {
+        courierApi.deleteCourier(courierApi.getIdCourier(CredentialsLoginPassword.from(courier)));
```
О!

pomegranate282922 commented on src/test/java/TestLoginCourier.java:
```diff
@@ -0,0 +1,63 @@
+        check.assertNotFoundRequest(response, "Учетная запись не найдена");
+    }
+    @AfterEach
+    public void terDown() {
```
tear

pomegranate282922 reviewed (approved):
Отлично, зачет!

Commit:
исправил опечатку.
